spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/rdd/partitioner/CassandraRDDPartitionGenerator.scala

```diff
@@ -0,0 +1,323 @@
+  *
+  * Under the hood uses a bound statement to generate routing keys which are then
+  * used the driver's internal token factories to determine the token for the
+  * routing key.
+  */
+private[connector] class CassandraRDDPartitioner[Key : ClassTag, V](
+  inputPartitions: Seq[CassandraPartition],
+  val indexedRanges: Seq[(Int, Seq[(Token[V], Token[V], Boolean)])],
+  val tableDef: TableDef,
+  connector: CassandraConnector,
+  tokenBounds: (Token[V], Token[V]))(
+implicit
+  @transient rwf:RowWriterFactory[Key]) extends Partitioner with Logging {
+
+  /**
+    * Adjust the partition key string in only of the Token clauses to be used with this scan
+    * incase the names differ from the inputPartitions
+    */
+  val partitions: Seq[CassandraPartition] = {
+    val pk = tableDef.partitionKey.map(_.columnName).map(quote).mkString(", ")
+    for (partition <- inputPartitions) yield {
+      val localTokenRanges = for (cqlTokenRange <- partition.tokenRanges) yield {
+        CqlTokenRange(cqlTokenRange.cql.changePartitionKey(pk), cqlTokenRange.values: _*)
+      }
+      partition.copy(tokenRanges = localTokenRanges)
+    }
+  }
+
+
+
+  def copy(
+    inputPartitions: Seq[CassandraPartition] = partitions,
+    indexedRanges: Seq[(Int, Seq[(Token[V], Token[V], Boolean)])] = indexedRanges,
+    tableDef: TableDef = tableDef,
+    connector: CassandraConnector = connector,
+    tokenBounds: (Token[V], Token[V]) = tokenBounds):CassandraRDDPartitioner[Key, V] = {
+
+    new CassandraRDDPartitioner[Key, V](partitions, indexedRanges, tableDef, connector, tokenBounds)
+  }
+
+  implicit val tO = tokenBounds._1.ord
+
+  val extractedPartitions = partitions
+    .map(cqlPartition => (cqlPartition.index, cqlPartition.tokenRanges))
+
+  val (minToken, maxToken) = tokenBounds
+  val minTokenValue = minToken.value
+  val maxTokenValue = maxToken.value
+
+  val partitionKeyWriter = implicitly[RowWriterFactory[Key]]
+      .rowWriter (tableDef, PartitionKeyColumns.selectFrom(tableDef))
+
+  /**
+   * Since the Token Generator relies on a (non-serializable) prepared statement we need to
+   * make sure it is not serialized to executors and is made fresh on each executor
+   */
+  @transient lazy val tokenGenerator = {
+    new TokenGenerator(connector, tableDef, partitionKeyWriter)
+  }
+
+  override def getPartition(key: Any): Int = {
+    key match {
+      case x: Key => {
+        val token = tokenGenerator.getTokenFor(x)
+        indexOfPartitionContaining(token)
+      }
+      case other => throw new IllegalArgumentException(s"Couldn't determine the key from object $other")
+    }
+  }
+
+  /**
+    * Since Driver TokenRange objects are not serializable we replicate the
+    * logic for TokenRange.contains(Token) here.
+    */
+  def indexOfPartitionContaining(token: com.datastax.driver.core.Token): Int = {
+    val tokenValue = token.getValue.asInstanceOf[V]
+    indexedRanges.find { case (index, ranges) =>
```

Comment:
Do I understand this right that this may get called for every key? If so, then it should be as cheap as possible. The number of tokens in the token ring may be very high when virtual nodes are enabled. You may be scanning several thousands of ranges here for bigger clusters.

I can see at least two better ways of finding the range:
1. build an interval tree (once) from ranges and then lookup a range in O(log n) time
2. use hashing-like technique: you can divide the token ring into 2^n buckets of equal size, close or little smaller than the average size of the token range and assign ranges to appropriate buckets. Because buckets are equal size, finding the bucket index for a token is O(1) and very fast (I guess a single bitwise shift-right) and then the remaining operation is just scanning a few ranges assigned to that bucket (which may be stored in a vector for faster scanning). 
Overall you'd get O(1) lookup and just two or three random memory accesses with this technique (think: nanoseconds fast).